src/bp/core/services/dialog/dialog-engine.ts

```diff
@@ -27,33 +32,122 @@ export class DialogEngine {
+    @inject(TYPES.InstructionProcessor) private instructionProcessor: InstructionProcessor,
+    @inject(TYPES.PromptManager) private promptManager: PromptManager,
+    @inject(TYPES.EventEngine) private eventEngine: EventEngine,
+    @inject(TYPES.DialogStore) private dialogStore: DialogStore
   ) {}
 
   public async processEvent(sessionId: string, event: IO.IncomingEvent): Promise<IO.IncomingEvent> {
     const botId = event.botId
     await this._loadFlows(botId)
 
-    const context = _.isEmpty(event.state.context) ? this.initializeContext(event) : event.state.context
-    const currentFlow = this._findFlow(botId, context.currentFlow)
-    const currentNode = this._findNode(botId, currentFlow, context.currentNode)
+    const context: IO.DialogContext = _.isEmpty(event.state.context)
+      ? this.initializeContext(event)
+      : event.state.context
+
+    const currentFlow = this._findFlow(botId, context.currentFlow!)
+    const currentNode = this._findNode(botId, currentFlow, context.currentNode!)
 
     if (event.ndu) {
       const workflowName = currentFlow.name?.replace('.flow.json', '')
 
       const { currentWorkflow } = event.state.session
       const { workflow } = event.state
 
-      if (currentWorkflow !== workflowName) {
+      if (currentWorkflow !== workflowName || !event.state.session.workflows?.[workflowName]) {
         this.changeWorkflow(event, workflowName)
         event.state.session.currentWorkflow = workflowName
-      }
+      } // TODO: this is weird, check
 
       const workflowEnded = currentNode.type === 'success' || currentNode.type === 'failure'
       if (workflowEnded && workflow) {
         workflow.success = currentNode.type === 'success'
         workflow.status = 'completed'
       }
+
+      if (currentNode.type === 'prompt' && !context.activePrompt && !this._getCurrentNodeValue(event, 'processed')) {
+        const { type, params } = currentNode.prompt!
+        context.activePrompt = {
+          stage: 'new',
+          status: 'pending',
+          state: {},
+          turn: 0,
+          config: {
+            type,
+            valueType: this.dialogStore.getPromptConfig(type)?.valueType,
+            ...params
+          }
+        }
+      }
+
+      if (context.activePrompt?.status === 'pending') {
+        const previousEvents =
+          context.activePrompt.stage === 'new'
+            ? await this._getPreviousEvents(event.target, context.activePrompt.config.searchBackCount)
+            : []
+
+        const { status: promptStatus, actions } = await this.promptManager.processPrompt(event, previousEvents)
+        context.activePrompt = promptStatus
+
+        for (const { type, payload, message, eventType } of actions) {
+          if (type === 'say') {
+            const incomingEventId = event.id
+
+            if (payload) {
+              await this.eventEngine.replyContentToEvent(payload, event, { incomingEventId, eventType })
```

Comment:
Actually replyContentToEvent is supposed to receive a content payload (eg: text, image, etc), but the payload is different. In that case, it sends the prompt configuration to the outgoing queue. The builtin module then intercepts that configuration, create a content payload (eg: type text) and send it again to the outgoing queue, for the channels to process. 

The "message" property is to keep the logic simple in prompt-manager, otherwise we'd have to keep repeating {type: 'text', text: ''}  (creating the whole payload) for each occurrences where we send a simple message